File: NeetCode/Heap/Last_Stone_Weight/last__stone__weight.py
from typing import Optional, List

import heapq


class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        pile = []
        for stone in stones:
            heapq.heappush(pile, -stone)

        while len(pile) > 1:
            first = -heapq.heappop(pile)
            second = -heapq.heappop(pile)

            if first == second:
                continue
            heapq.heappush(pile, -abs(first - second))

        if len(pile) == 0:
            return 0
        return abs(pile[0])

        pass


if __name__ == "__main__":
    sol = Solution()
    stones = [3, 7, 2]

    print(sol.lastStoneWeight(stones))
    print("Running Solution...")

File: NeetCode/Heap/Last_Stone_Weight/Last_Stone_Weight.cpp
// Last_Stone_Weight.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    int lastStoneWeight(vector<int> &stones)
    {
        priority_queue<int> pile;
        for (int stone : stones)
        {
            pile.push(stone);
        }
        while (pile.size() > 1)
        {
            int first = pile.top();
            pile.pop();
            int second = pile.top();
            pile.pop();
            if (first == second)
            {
                continue;
            }
            pile.push(abs(first - second));
        }
        if (pile.size() == 0)
        {
            return 0;
        }
        return pile.top();
    }
};
